src/generate_csv.py

```diff
@@ -0,0 +1,86 @@
+import boto3
+from .utils import ensure, first, splitfilter, select_keys, keys
+import time
+import sys
+import os
+import csv
+
+def client():
+    return boto3.client('iam')
+
+def coerce(row):
+    value_lookups = {
+        'false': False,
+        'true': True,
+        'N/A': None,
+        'no_information': None # wtf? treat as N/A
+    }
+    return {key: value_lookups.get(val, val) for key, val in row.items()}
+
+def load_csv(path):
+    with open(path, 'r') as csvfile:
+        reader = csv.DictReader(csvfile)
+        return [coerce(row) for row in reader]
+
+def dump_csv(path, list_of_dicts):
+    with open(path, 'w') as csvfile:
+        fieldnames = keys(first(list_of_dicts))
+        writer = csv.DictWriter(csvfile, fieldnames)
+        writer.writeheader()
+        for row in list_of_dicts:
+            writer.writerow(row)
+    return path
+
+def num_uppercase(string):
+    return sum(1 for char in string if char.isupper())
+
+# ---
+
+def generate_credential_report():
+    iam = client()
+    print('requesting report')
+    while True:
+        sys.stdout.write('polling ... ')
+        resp = iam.generate_credential_report()
+        if resp not in ['STARTED', 'INPROGRESS']:
+            print('done.')
+            break
+        time.sleep(2) # seconds
+    ensure(resp['State'] == 'COMPLETE', "failed to generate credential report. final response: %s" % resp)
+    resp = iam.get_credential_report()
+    ensure(resp['ResponseMetadata']['HTTPStatusCode'] == 200, "failed to download credential report. final response: %s" % resp)
+    ensure(resp['ReportFormat'] == 'text/csv', "unexpected report format %r. final response: %s" % (resp['ReportFormat'], resp))
+    filename = 'aws-credentials-report.private.csv'
+    open(filename, 'wb').write(resp['Content'])
+    print("wrote 'aws-credentials-report.private.csv'")
+    return filename
+
+def target_humans(path_to_credentials_report):
+    """interesting humans have either access keys 1 or 2 active, or a password enabled, and at least two uppercase letters in their name.
+    The rest are MACHINES or candidates for deletion"""
+    ensure(os.path.exists(path_to_credentials_report), "credentials report does not exist")
+    report = load_csv(path_to_credentials_report)
+
+    # humans with machine names
+    exceptions = [
+        '<root_account>',
+        'nathanlisgo',
+        'james_gilbert',
+        'melissa_harrison'
```

Comment:
can't rename this, though we could create new users and transition them

Reply:
that's a good idea, I hadn't thought of that. I don't like the idea of having people's names in this code either.

Reply:
I think `nathanlisgo` might still be an exception. Probably can't have both `nathanlisgo` and `NathanLisgo`. I'll try it ..

Reply:
yeah, nah.